test/json_unittest.cc

```diff
@@ -0,0 +1,428 @@
+      {"baz", json::boolean(true)},
+      {"str", json::string("asdfasdf")},
+    });
+    EXPECT_TOSTRING_EQ(
+      "{"
+      "\"bar\":{\"x\":0.0,\"y\":null},"
+      "\"baz\":true,"
+      "\"foo\":[1.0,2.0,3.0],"
+      "\"str\":\"asdfasdf\""
+      "}",
+      v);
+  });
+}
+
+TEST(BigTest, LotsOfArrayNesting) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString(
+      "[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[["
+      "]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]");
+    EXPECT_TOSTRING_EQ(
+      "[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[["
+      "]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]",
+      v);
+  });
+}
+
+TEST(BigTest, LotsOfObjectNesting) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString(
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":"
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":"
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":"
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":null"
+      "}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}");
+    EXPECT_TOSTRING_EQ(
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":"
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":"
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":"
+      "{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":{\"f\":null"
+      "}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}",
+      v);
+  });
+}
+
+
+// Parse errors.
+
+TEST(ParseError, Empty) {
+  ASSERT_THROW(json::Parser::FromString(""), json::Exception);
+}
+
+TEST(ParseError, UnexpectedChar) {
+  ASSERT_THROW(json::Parser::FromString("x"), json::Exception);
+}
+
+TEST(ParseError, UnterminatedArray) {
+  ASSERT_THROW(json::Parser::FromString("["), json::Exception);
+}
+
+TEST(ParseError, UnmatchedArrayClose) {
+  ASSERT_THROW(json::Parser::FromString("]"), json::Exception);
+}
+
+TEST(ParseError, UnterminatedObject) {
+  ASSERT_THROW(json::Parser::FromString("{"), json::Exception);
+}
+
+TEST(ParseError, UnmatchedObjectClose) {
+  ASSERT_THROW(json::Parser::FromString("}"), json::Exception);
+}
+
+TEST(ParseError, UnterminatedString) {
+  ASSERT_THROW(json::Parser::FromString("\""), json::Exception);
+}
+
+TEST(ParseError, UnterminatedEscape) {
+  ASSERT_THROW(json::Parser::FromString("\"\\\""), json::Exception);
+}
+
+TEST(ParseError, UnterminatedUnicodeEscape) {
+  ASSERT_THROW(json::Parser::FromString("\"\\u\""), json::Exception);
+  ASSERT_THROW(json::Parser::FromString("\"\\u0\""), json::Exception);
+  ASSERT_THROW(json::Parser::FromString("\"\\u00\""), json::Exception);
+  ASSERT_THROW(json::Parser::FromString("\"\\u000\""), json::Exception);
+}
+
+TEST(ParseError, ObjectNoValue) {
+  ASSERT_THROW(json::Parser::FromString("{\"x\"}"), json::Exception);
+  ASSERT_THROW(json::Parser::FromString("{\"x\":}"), json::Exception);
+}
+
+
```

Comment:
You can remove one of these two blank lines.

Reply:
Done.